src/modules/linux-micro/automount/automount.c

```diff
@@ -0,0 +1,347 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "linux-micro-automount");
+
+#include "sol-file-reader.h"
+#include "sol-mainloop.h"
+#include "sol-platform-linux-micro.h"
+#include "sol-platform.h"
+#include "sol-util.h"
+
+#define EXTMAGIC "\123\357\001"
+#define EXT_SB_OFFSET 1024
+#define EXT_MAGIC_OFFSET EXT_SB_OFFSET + 0x38
+#define EXT_FEATURE_OFFSET EXT_SB_OFFSET + 0x5c
+
+#define EXT3_FEATURE_COMPAT_HAS_JOURNAL     0x0004
+
+#define EXT2_FEATURE_RO_COMPAT_SPARSE_SUPER 0x0001
+#define EXT2_FEATURE_RO_COMPAT_LARGE_FILE   0x0002
+#define EXT2_FEATURE_RO_COMPAT_BTREE_DIR    0x0004
+
+#define EXT2_FEATURE_INCOMPAT_FILETYPE      0x0002
+#define EXT3_FEATURE_INCOMPAT_RECOVER       0x0004
+#define EXT3_FEATURE_INCOMPAT_JOURNAL_DEV   0x0008
+#define EXT2_FEATURE_INCOMPAT_META_BG       0x0010
+
+#define EXT2_FEATURE_RO_COMPAT_SUPP (EXT2_FEATURE_RO_COMPAT_SPARSE_SUPER | \
+    EXT2_FEATURE_RO_COMPAT_LARGE_FILE | \
+    EXT2_FEATURE_RO_COMPAT_BTREE_DIR)
+#define EXT2_FEATURE_INCOMPAT_SUPP  (EXT2_FEATURE_INCOMPAT_FILETYPE | \
+    EXT2_FEATURE_INCOMPAT_META_BG)
+#define EXT2_FEATURE_INCOMPAT_UNSUPPORTED   ~EXT2_FEATURE_INCOMPAT_SUPP
+#define EXT2_FEATURE_RO_COMPAT_UNSUPPORTED  ~EXT2_FEATURE_RO_COMPAT_SUPP
+
+#define EXT3_FEATURE_RO_COMPAT_SUPP (EXT2_FEATURE_RO_COMPAT_SPARSE_SUPER | \
+    EXT2_FEATURE_RO_COMPAT_LARGE_FILE | \
+    EXT2_FEATURE_RO_COMPAT_BTREE_DIR)
+#define EXT3_FEATURE_INCOMPAT_SUPP  (EXT2_FEATURE_INCOMPAT_FILETYPE | \
+    EXT3_FEATURE_INCOMPAT_RECOVER | \
+    EXT2_FEATURE_INCOMPAT_META_BG)
+#define EXT3_FEATURE_INCOMPAT_UNSUPPORTED   ~EXT3_FEATURE_INCOMPAT_SUPP
+#define EXT3_FEATURE_RO_COMPAT_UNSUPPORTED  ~EXT3_FEATURE_RO_COMPAT_SUPP
+
+/** superblock is the same for ext2/3/4 **/
+struct ext_super_block {
+    uint32_t s_feature_compat;
+    uint32_t s_feature_incompat;
+    uint32_t s_feature_ro_compat;
+} __attribute__((packed));
+
+struct filesystem_magic {
+    const char *magic;
+    int length;
+    int offset;
+};
+
+struct filesystem {
+    const char *id;
+    const struct filesystem_magic *magics;
+    bool (*fs_cb)(int fd);
+};
+
+static const struct filesystem_magic vfat_magic[] = {
+    { .magic = "MSWIN",    .length = 5, .offset = 0x52 },
```

Comment:
Perhaps a macro here will help maintain the table?

    #define MAGIC(str_, offset_)  { .magic = (str_), .length = sizeof(str_) - 1,  .offset = (offset_) }
        MAGIC("MSWIN", 0x52),
        MAGIC("FAT32   ", 0x52),
        MAGIC("MSDOS", 0x36),
        ...
    #undef MAGIC